### ISSUE TITLE
Neverloop false positive on try blocks

### ISSUE THREAD

**Issue by AstrickHarren** [rust-lang/rust-clippy#12205]
### Summary

see title

### Lint Name

clippy::never_loop

### Reproducer

I tried this code:

```rust 
fn test() -> Option<()> {
    loop {
        let _: Option<_> = try {
            println!("looping");
            None?;
            return Some(());
        };
    }
}
```

I saw this happen:

```rust
warning: this loop never actually loops
  --> src/main.rs:21:5
   |
21 | /     loop {
22 | |         let _: Option<_> = try {
23 | |             println!("looping");
24 | |             None?;
25 | |             return Some(());
26 | |         };
27 | |     }
   | |_____^
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#never_loop
   = note: `#[warn(clippy::never_loop)]` on by default

```

I expected to see this happen:
```rust
    Finished dev [unoptimized + debuginfo] target(s) in 0.87s
     Running `target/debug/foo`

looping
looping
looping
...
```

### Version

```text
rustc 1.77.0-nightly (88189a71e 2024-01-19)
binary: rustc
commit-hash: 88189a71e4e4376eea82ac61db6a539612eb200a
commit-date: 2024-01-19
host: x86_64-unknown-linux-gnu
release: 1.77.0-nightly
LLVM version: 17.0.6
```
